atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogReader.java

```diff
@@ -16,100 +15,49 @@
-      final RaftLog log, final JournalReader journalReader, final RaftLogReader.Mode mode) {
-    this.log = log;
-    this.journalReader = journalReader;
-    this.mode = mode;
-
-    nextIndex = log.getFirstIndex();
-  }
-
-  @Override
-  public boolean hasNext() {
-    return (mode == Mode.ALL || nextIndex <= log.getCommitIndex()) && journalReader.hasNext();
-  }
-
-  @Override
-  public IndexedRaftLogEntry next() {
-    if (!hasNext()) {
-      throw new NoSuchElementException();
-    }
-
-    final JournalRecord journalRecord = journalReader.next();
-    final RaftLogEntry entry = serializer.readRaftLogEntry(journalRecord.data());
-
-    nextIndex = journalRecord.index() + 1;
-    return new IndexedRaftLogEntryImpl(entry.term(), entry.entry(), journalRecord);
-  }
-
-  public long reset() {
-    nextIndex = journalReader.seekToFirst();
-    return nextIndex;
-  }
-
-  public long seek(final long index) {
-    long boundedIndex = index;
-
-    if (mode == Mode.COMMITS) {
-      // allow seeking one past the commit index to simulate being at the end of the log
-      final long upperBoundIndex = log.getCommitIndex() + 1;
-      boundedIndex = Math.min(index, upperBoundIndex);
-    }
-
-    nextIndex = journalReader.seek(boundedIndex);
-    return nextIndex;
-  }
-
-  public long seekToLast() {
-    if (mode == Mode.ALL) {
-      nextIndex = journalReader.seekToLast();
-    } else {
-      seek(log.getCommitIndex());
-    }
-
-    return nextIndex;
-  }
-
-  public long seekToAsqn(final long asqn) {
-    if (mode == Mode.COMMITS) {
-      nextIndex = journalReader.seekToAsqn(asqn, log.getCommitIndex());
-    } else {
-      nextIndex = journalReader.seekToAsqn(asqn);
-    }
-    return nextIndex;
-  }
+public interface RaftLogReader extends java.util.Iterator<IndexedRaftLogEntry>, AutoCloseable {
+
+  /**
+   * Reset the reader to the first index of the log.
+   *
+   * @return return the first index
+   */
+  long reset();
+
+  /**
+   * Seeks to the given index.
+   *
+   * @param index the index to seeks to
+   * @return the index of the next record to be read
+   */
+  long seek(long index);
+
+  /**
+   * Seeks to the last index
+   *
+   * @return the index of the next record to be read
+   */
+  long seekToLast();
+
+  /**
+   * Seek to a record with the highest ASQN less than or equal to the given asqn.
+   *
+   * @param asqn the asqn to seek to
+   * @return the index of the record that will be returned by {@link #next()}
+   */
+  long seekToAsqn(final long asqn);
+
+  /**
+   * Seek to a record with the highest ASQN less than or equal to the given asqn, with the
+   * restriction that it seeks to a record with index less than or equal to the given
+   * indexUpperBound.
+   *
+   * @param asqn the asqn to seek to
+   * @param indexUpperBound the index until which it seeks
+   * @return the index of the record that will be returned by {@link #next()}
+   */
+  long seekToAsqn(final long asqn, final long indexUpperBound);
```

Comment:
:wrench: This method could be dropped. It is only called by `RaftCommittedReader`. Unless we anticipate a use for it in the future